File: src/main/java/com/wooteco/wiki/document/service/DocumentSearchService.java
package com.wooteco.wiki.document.service;

import com.wooteco.wiki.document.domain.dto.DocumentSearchResponse;
import com.wooteco.wiki.document.repository.DocumentRepository;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@RequiredArgsConstructor
@Service
@Transactional
public class DocumentSearchService {

    private final DocumentRepository documentRepository;

    @Transactional(readOnly = true)
    public List<DocumentSearchResponse> search(String keyWord) {
        return documentRepository.findAllByTitleStartingWithOrderByTitle(keyWord)
                .stream()
                .map(document -> new DocumentSearchResponse(
                        document.getTitle(),
                        document.getUuid(),
                        document.type()
                ))
                .toList();
    }
}

File: src/main/java/com/wooteco/wiki/document/service/DocumentSearchService.kt
package com.wooteco.wiki.document.service

import com.wooteco.wiki.document.domain.dto.DocumentSearchResponse
import com.wooteco.wiki.document.repository.DocumentRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional
class DocumentSearchService(
    private val documentRepository: DocumentRepository,
) {

    @Transactional(readOnly = true)
    fun search(keyWord: String): List<DocumentSearchResponse> =
        documentRepository.findAllByTitleStartingWith(keyWord)
            .map { DocumentSearchResponse(it.title, it.uuid) }
}
